lib/Sema/CSSimplify.cpp

```diff
@@ -9853,8 +9862,48 @@ ConstraintSystem::simplifyKeyPathConstraint(
 
   auto tryMatchRootAndValueFromType = [&](Type type,
                                           bool allowPartial = true) -> bool {
-    Type boundRoot = Type(), boundValue = Type();
 
+    // If we have a function type, attempt an SE-0249 key-path-function
+    // conversion.
+    if (auto fnTy = type->getAs<FunctionType>()) {
+      if (fnTy->getParams().size() != 1) {
+        if (!shouldAttemptFixes())
+          return false;
+
+        resolveAsMultiArgFuncFix = true;
+        auto *fix = AllowMultiArgFuncKeyPathMismatch::create(
+            *this, fnTy, locator.getBaseLocator());
+        // Pretend the keypath type got resolved and move on.
+        return !recordFix(fix);
+      }
+
+      definitelyFunctionType = true;
+
+      // Key paths may be converted to a function of compatible type. We will
+      // later form from this key path an implicit closure of the form
+      // `{ root in root[keyPath: kp] }` so any conversions that are valid with
+      // a source type of `(Root) -> Value` should be valid here too.
+      auto rootParam = AnyFunctionType::Param(rootTy);
+      auto kpFnTy = FunctionType::get(rootParam, valueTy, fnTy->getExtInfo());
+
+      // Note: because the keypath is applied to `root` as a parameter internal
+      // to the closure, we use the function parameter's "parameter type" rather
+      // than the raw type. This enables things like:
+      // ```
+      // let countKeyPath: (String...) -> Int = \.count
```

Comment:
Could you please add this as a test-case as well? Also we probably want to test for `inout`.

Reply:
We have test cases for variadic params already, but I'll add some for `inout`!